### PR TITLE
[v1.8] NCL-5527 - Update lastModificationTime only if audited fields are updated

### PR DESCRIPTION
**Pull request by vibe13**
### Checklist:

* [ ] Have you added a note in the [CHANGELOG wiki](https://github.com/project-ncl/pnc/wiki/Changelog) for your change if user-facing?
* [ ] Have you added unit tests for your change?

### CODE REVIEW

File: datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationRepositoryImpl.java [hunk 1]
```diff
@@ -53,15 +54,51 @@ public BuildConfiguration save(BuildConfiguration buildConfiguration) {
...
+                equalsWithNull(persisted.getBuildScript(), toUpdate.getBuildScript()) &&
+                equalsId(persisted.getRepositoryConfiguration(), toUpdate.getRepositoryConfiguration()) &&
+                equalsWithNull(persisted.getScmRevision(), toUpdate.getScmRevision()) &&
+                equalsId(persisted.getProject(), toUpdate.getProject()) &&
+                equalsId(persisted.getBuildEnvironment(), toUpdate.getBuildEnvironment()) &&
+                (persisted.isArchived() == toUpdate.isArchived()) &&
+                (persisted.getBuildType() == toUpdate.getBuildType());
+    }
+
+    private boolean equalsWithNull(Object o1, Object o2) {
```
Comment (janinko) [inline datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationRepositoryImpl.java]: reinwenting wheel: https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#equals-java.lang.Object-java.lang.Object-

File: datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationRepositoryImpl.java [hunk 2]
```diff
@@ -53,15 +54,51 @@ public BuildConfiguration save(BuildConfiguration buildConfiguration) {
...
+            return true;
+        }
+
+        if (persisted == null && toUpdate != null) {
+            return false;
+        }
+
+        if (persisted != null && toUpdate == null) {
+            return false;
+        }
```
Comment (janinko) [inline datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationRepositoryImpl.java]: You can shorten it as:
```suggestion
        if (persisted == null || toUpdate == null) {
            return false;
        }
```
Not sure which one is better. For mine you need to consider previous guard, yours is more explicit.

File: integration-test/src/test/java/org/jboss/pnc/integration/BuildTest.java [hunk 3]
```diff
@@ -264,25 +270,103 @@ public void shouldBuildTemporaryBuildAndNotAssignItToMilestone() {
...
+        RestResponse<BuildRecordRest> forcedPersistentBuild = triggerBCBuild(buildConfigurationRest, Optional.empty(), persistent);
+        assertThat(forcedPersistentBuild.getRestCallResponse().getStatusCode()).isEqualTo(200);
+        ResponseUtils.waitSynchronouslyFor(() -> {
+            RestResponse<BuildRecordRest> record = buildRecordRestClient.get(forcedPersistentBuild.getValue().getId(),
+                    false);
+            return record.hasValue() && record.getValue().getStatus().equals(BuildCoordinationStatus.DONE);
+        } , 15, TimeUnit.SECONDS);
+
+        // Update only description, should not create a new revision and keep same lastModificationTime
+        String oldDescription = buildConfigurationRest.getDescription() != null ? new String(buildConfigurationRest.getDescription()) : null;
```
Comment (janinko) [inline integration-test/src/test/java/org/jboss/pnc/integration/BuildTest.java]: why do you call `new String` here?

Comment (janinko) [reply integration-test/src/test/java/org/jboss/pnc/integration/BuildTest.java]: Why isn't this enough? Strings are immutable
```suggestion
        String oldDescription = buildConfigurationRest.getDescription();
```

File: integration-test/src/test/java/org/jboss/pnc/integration/BuildTest.java [hunk 4]
```diff
@@ -264,25 +270,103 @@ public void shouldBuildTemporaryBuildAndNotAssignItToMilestone() {
...
+        Date oldLastModDate = buildConfigurationRest.getLastModificationTime() != null ? new Date(buildConfigurationRest.getLastModificationTime().getTime()) : null;
```
Comment (janinko) [inline integration-test/src/test/java/org/jboss/pnc/integration/BuildTest.java]: The same here. I know Date is mutable, but is the lastModificationTime being altered?

File: datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationRepositoryImpl.java (hunk 1 above)
Comment (vibe13) [reply datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationRepositoryImpl.java]: Ah thanks, ok using existing wheel now :-), here and in another place where I first found it

File: datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationRepositoryImpl.java (hunk 2 above)
Comment (vibe13) [reply datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationRepositoryImpl.java]: I applied yours as it's better, here an in another method in same class

File: integration-test/src/test/java/org/jboss/pnc/integration/BuildTest.java (hunk 3 above)
Comment (vibe13) [reply integration-test/src/test/java/org/jboss/pnc/integration/BuildTest.java]: They are enough, this is leftover of junk code I put when I was frustrated by tests not passing (due to other issues), which I did not remove. Thanks!

File: integration-test/src/test/java/org/jboss/pnc/integration/BuildTest.java (hunk 4 above)
Comment (vibe13) [reply integration-test/src/test/java/org/jboss/pnc/integration/BuildTest.java]: Same as above

**Comment by vibe13** [comment]
@janinko is it all good now? Thanks! :-)

**Pull request merged**